src/core/meshcop/dataset_local.cpp

```diff
@@ -270,26 +270,57 @@ void DatasetLocal::EmplaceSecurelyStoredKeys(Dataset &aDataset) const
     KeyRef         pskcRef       = IsActive() ? kActiveDatasetPskcRef : kPendingDatasetPskcRef;
     NetworkKeyTlv *networkKeyTlv = aDataset.GetTlv<NetworkKeyTlv>();
     PskcTlv *      pskcTlv       = aDataset.GetTlv<PskcTlv>();
+    bool           moveKeys      = false;
     size_t         keyLen;
+    Error          error;
 
     if (networkKeyTlv != nullptr)
     {
         // If the dataset contains a network key, its real value must have been moved to
         // the secure storage upon saving the dataset, so restore it back now.
         NetworkKey networkKey;
-        SuccessOrAssert(ExportKey(networkKeyRef, networkKey.m8, NetworkKey::kSize, keyLen));
-        OT_ASSERT(keyLen == NetworkKey::kSize);
-        networkKeyTlv->SetNetworkKey(networkKey);
+        error = ExportKey(networkKeyRef, networkKey.m8, NetworkKey::kSize, keyLen);
```

Comment:
Now I think it's a pity that we don't differentiate between "key doesn't exist" and other types of errors because we can't properly detect other errors here. Maybe something to think about :)

Reply:
I agree. We need to extend the key management APIs to return more detailed error codes, right now it is either `kErrorInvalidArgs` or `kErrorFailed`. But let us take that up in a separate PR.